modules/custom/openy_focal_point/src/Plugin/Field/FieldWidget/OpenYFocalPointImageWidget.php

```diff
@@ -0,0 +1,225 @@
+        ]),
+      '#attached' => [
+        'library' => ['core/drupal.dialog.ajax'],
+      ],
+      '#attributes' => [
+        'class' => ['use-ajax'],
+        'data-selector' => $element_selectors['focal_point'],
+        'data-field-name' => $field_name,
+        'data-dialog-type' => 'modal',
+        'target' => '_blank',
+      ],
+    ];
+
+    return $preview_link;
+  }
+
+  private static function createFocalPointEditLink($fid, $field_name, array $element_selectors, $default_focal_point_value, $image_styles) {
+    // Replace comma (,) with an x to make javascript handling easier.
+    $preview_focal_point_value = str_replace(',', 'x', $default_focal_point_value);
+
+    // Create a token to be used during an access check on the preview page.
+    $token = self::getPreviewToken();
+
+    $preview_link = [
+      '#type' => 'link',
+      '#title' => new TranslatableMarkup('Set Focal Point'),
+      '#url' => new Url('openy_focal_point.edit_focal_point',
+        [
+          'fid' => $fid,
+          'focal_point_value' => $preview_focal_point_value,
+          'field_name' => $field_name,
+        ],
+        [
+          'query' => ['focal_point_token' => $token, 'image_styles' => implode(':', $image_styles)],
+        ]),
+      '#attached' => [
+        'library' => ['core/drupal.dialog.ajax'],
+      ],
+      '#attributes' => [
+        'class' => ['use-ajax'],
+        'data-selector' => $element_selectors['focal_point'],
+        'data-field-name' => $field_name,
+        'data-dialog-type' => 'modal',
+        'target' => '_blank',
+      ],
+    ];
+
+    return $preview_link;
+  }
+
+  /**
+   * Loads the entity view display entity and prepare list of all image styles
+   * used (to be passed with preview link).
+   */
+  public static function process($element, FormStateInterface $form_state, $form) {
+    $element = parent::process($element, $form_state, $form);
+
+    if (isset($element['focal_point'])) {
+      $element['focal_point']['#access'] = FALSE;
+      $element['preview']['indicator']['#access'] = FALSE;
+    }
+
+    if (isset($element['preview']['preview_link'])) {
+      $item = $element['#value'];
+      $fid = $item['fids'][0];
+      $element_selectors = [
+        'focal_point' => 'focal-point-' . implode('-', $element['#parents']),
+      ];
+      $default_focal_point_value = isset($item['focal_point']) ? $item['focal_point'] : $element['#focal_point']['offsets'];
+
+      // Search through $form_state['input'] for target_id => 'media:XX' element. If found
+      // use "entity_browser_widget" right next to it to get id that can be used
+      // to load image styles for the paragraph.
+      $paragraph_type = self::getParagraphInfo($fid, $form_state);
+
+      // Another big assumption here. We assume that Media has view modes named
+      // "prgf_<paragraph_name>".
+      $display = \Drupal::entityTypeManager()->getStorage('entity_view_display')->load('media.image.prgf_' . $paragraph_type);
```

Comment:
we should pass service via DI

Reply:
we use this service in public static method, so we don't have instance of the form. This is why we can not use DI here.